proposals/ref-lifetimes.md

```diff
@@ -0,0 +1,294 @@
+
+As it is, as long as we can express the C# lifetime conventions in the above syntax in a way
+which type checks (according to conventional rules of generic substition and compatibility), safety
+is guaranteed.
+
+However, we must also account for ref structs and ref fields. In fact, the problem of ref structs
+is almost fully analogous to methods. Ref fields, like ref parameters, do not have a single lifetime.
+Instead, the lifetime is defined at the point of creation (or substitution). However, because ref 
+structs do not end their life at the end of their constructor, they must themselves have a lifetime.
+According to the basic rule of lifetime safety, the lifetime of a ref must be shorter than its target,
+and therefore the lifetime of a ref struct must be the shortest of all the lifetime parameters passed.
+
+We can use a similar syntax extension for ref structs and ref fields as well:
+
+```C#
+ref struct S<'a, 'b, T>
+{
+    ref<'a> T F1;
+    ref<'b> int F2;
+}
+```
+
+As with all generic variables, they must be compatible to support assignment. As stated before, only longer
+lifetimes are assignable to shorter lifetimes. Notably, lifetimes like `'a` and `'b` provided above are
+not comparable as given and therefore cannot be assigned to one another.
+
+For example:
+
+```C#
+S<'a, 'b, T> M<'a, 'b, T>(ref<'a> T p1, ref<'b> int p2)
+{
+    var s1 = new S<'a, 'b, T>() { F1 = p1, F2 = p2 }; // OK, type checks as equal lifetimes
+    return s1; // OK, types match
+
+    int x = 0;
+    var s2 = new S<'a, 'b, T>() { F1 = p1, F2 = ref x }; // Error, `x` is method-local lifetime, 'b is longer
+    return s2; // OK, types match
+
+    var s3 = new S<'a, 'b, T>() { F1 = p1, F2 = ref (new int[] { 2 })[0] }; // OK, arrays have global lifetime and can be assigned to any variable
+    return s3; // OK types match
+
+    var s4 = new S<T>() { F1 = p1, F2 = ref x }; // OK, lifetime of S.F2 is inferred as method-local lifetime
+    return s4; // Error: s4 lifetime is method-local, shorter than 'a & 'b
+}
+```
+
+## Syntax simplifactions
+
+### Implicit lifetimes
+
+While the above construction is very flexible and can represent almost any lifetime construction existing
+or being considered for C#, it's worthwhile to consider other syntaxes and how they could map to the above.
+The main advantage of this is that we can reduce the safety verification to a question of whether the
+proposed translation type checks according to the above rules. If it does, and the above rules are safe, then
+the alternate syntax is sound.
+
+First, let's consider existing C#, which doesn't have any lifetime annotations. We can work out the implicit
+annotations by examining the existing C# safety rules.
+
+For the simplest cases, only one ref parameter or return type, there is only one possible translation:
+
+```C#
+// Case 1
+ref int M1() { ... }
+// Case 2
+void M2(ref int x) { ...}
+// Case 3
+Span<int> M3() { ... }
+// Case 4
+void M4<'a>(Span<'a, int> s) { ... }
```

Comment:
Is this a typo? Case 4 maps to case 4? (Not sure if I'm using this Github feature properly, let's see.)